### PR TITLE
Finalized should be before best

### PR DESCRIPTION
**Pull request by davxy**
In current implementation of `apply_finality_with_block_hash` we explicitly check the `finalized < best` condition
only if there is more than one leaf (in this case because `finalized` may be in a different branch than last `best`.

If this is the case then we explicitly set the newly finalized block as the best.

(as the comment says) This condition "weak check" is due to the assumption that the best is always after the finalized because consensus algorithms typically guarantees this.

But this is not always the case.

For example, in cumulus parachain code we are maintaining a cache of parablocks that (according to the relay-chain) are finalized. All these blocks are explicitly finalized via the `Finalized::finalize_block` without worrying about updating the best first [here](https://github.com/paritytech/cumulus/blob/5b0db49b5c2ed417d3673ecc80910cf92f103bf1/client/consensus/common/src/parachain_consensus.rs#L141)

---

This PR adds a defensive check that enforces the condition and thus handles correctly finalizations that are applied without updating `best` first.

---

This PR is necessary for [THIS](https://github.com/paritytech/cumulus/pull/2696) cumulus PR to work correctly

### CODE REVIEW

File: client/service/src/client/client.rs [hunk 1]
```diff
@@ -907,52 +907,56 @@ where
...
 			);
 
 			return Err(sp_blockchain::Error::NotInFinalizedChain)
 		}
 
-		// If there is only one leaf, best block is guaranteed to be
-		// a descendant of the new finalized block. If not,
-		// we need to check.
-		if self.backend.blockchain().leaves()?.len() > 1 {
+		// We may need to cohercively update the best block if there is more than one
```
Comment (koute) [inline client/service/src/client/client.rs:937]:
```suggestion
		// We may need to coercively update the best block if there is more than one
```

File: client/service/src/client/client.rs [hunk 2]
```diff
@@ -907,52 +907,56 @@ where
...
+		// leaf or if the finalized block number is greater than last best number recorded
+		// by the backend. This last condition may apply in case of consensus implementations
+		// not always checking this condition.
+		let block_number = self
+			.backend
+			.blockchain()
+			.number(hash)?
+			.ok_or(Error::MissingHeader(format!("{hash:?}")))?;
+		if self.backend.blockchain().leaves()?.len() > 1 || info.best_number < block_number {
```
Comment (skunert) [inline client/service/src/client/client.rs:946]: Hmm this new check only triggers when there is only one leaf. In that case, we could directly mark_head since there is no other ancestry possible, no need to check the tree route? Even though impact is probably super low since these headers are probably cached.

**Review by koute — COMMENTED** [review]
This seems like a tricky corner case; do we have a test which verifies this behavior? If not can we add one?

**Review by skunert — APPROVED** [review]

File: client/service/src/client/client.rs [hunk 3]
```diff
@@ -907,52 +907,56 @@ where
...
+		// We may need to coercively update the best block if there is more than one
+		// leaf or if the finalized block number is greater than last best number recorded
+		// by the backend. This last condition may apply in case of consensus implementations
+		// not always checking this condition.
+		let block_number = self
+			.backend
+			.blockchain()
+			.number(hash)?
+			.ok_or(Error::MissingHeader(format!("{hash:?}")))?;
+		if self.backend.blockchain().leaves()?.len() > 1 || info.best_number < block_number {
```
Comment (bkchr) [inline client/service/src/client/client.rs:946]: Some small wonderful test maybe? :see_no_evil:

**Review by bkchr — APPROVED** [review]
